File: 42-trapping-rain-water/42-trapping-rain-water.ts
function trap(height: number[]): number {
    var arraylength: number = height.length
    var leftPointer: number = 0
    var leftMax: number = height[leftPointer]
    var rightPointer: number = arraylength - 1
    var rightMax: number = height[rightPointer]
    var ans: number = 0;
    var currentIndex = 0
    while (leftPointer < rightPointer) {

        let min: number = leftMax < rightMax ? leftMax : rightMax;
        let currentValue = min - height[currentIndex]
        ans = currentValue > 0 ? ans + currentValue : ans;

        if (height[leftPointer] < height[rightPointer]) {
            leftPointer++;
            leftMax = leftMax > height[leftPointer] ? leftMax : height[leftPointer];
            currentIndex = leftPointer
        }
        else {
            rightPointer--;
            rightMax = rightMax > height[rightPointer] ? rightMax : height[rightPointer];
            currentIndex = rightPointer
        }
    }
    return ans;
};

File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) == 0:
            return 0
        left_index, right_index = 0, len(height) - 1
        left_max, right_max = height[left_index], height[right_index]
        ans = 0
        while left_index < right_index:
            if left_max <= right_max:
                temp = left_max - height[left_index]
                if temp > 0:
                    ans += temp
                left_index += 1
                left_max = max(left_max, height[left_index])
                continue
            else:
                temp = right_max - height[right_index]
                if temp > 0:
                    ans += temp
                right_index -= 1
                right_max = max(right_max, height[right_index])
                continue
        return ans